// yi-frames-ui/utils/src/js/common/mods/gotop.js
export let goTop = function (ele) {
    let goTopTem;
    if (ele) {
        goTopTem = ele
    } else {
        const topDom = document.createElement("a")
        topDom.href = 'javascript:;'
        topDom.className = 'yi-gotop'
        document.body.appendChild(topDom)
        goTopTem = document.querySelector('.yi-gotop')
    }
    new class ToBack {
        constructor(obj, time = 800) {
            this.obj = obj
            this.time = time;
            this.bOk = false;
            this.init();
        }
        init() {
            let _this = this;
            window.onscroll = () => {
                if (_this.bOk) {
                    clearInterval(_this.obj.timer)
                }
                _this.bOk = true;
                _this.showHide();
            };
            this.obj.onclick = () => {
                _this.scroll();
            }
        }
        showHide() {
            let scrollTop = document.documentElement.scrollTop || document.body.scrollTop;
            if (scrollTop > 40) {
                this.obj.classList.add("block");
            } else {
                this.obj.classList.remove("block");
            }
        }
        scroll() {
            let start = document.documentElement.scrollTop || document.body.scrollTop,
                iTarget = 0,
                dis = iTarget - start,
                count = Math.floor(this.time / 50),
                n = 0,
                that = this;
            clearInterval(this.obj.timer);
            this.obj.timer = setInterval(function () {
                //js滚动的时候
                that.bOk = false;
                n++;
                let a = 1 - n / count;
                document.documentElement.scrollTop = document.body.scrollTop = start + dis * (1 - Math.pow(a, 3));
                if (n == count) {
                    clearInterval(that.obj.timer);
                }
            }, 30);
        }
    }(goTopTem)
}
